File: android/app/src/main/java/com/gang/notodo/calendar/IndexMonthView.java
package com.gang.notodo.calendar;


import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;

import com.gang.notodo.util.CalendarUtil;
import com.haibin.calendarview.Calendar;
import com.haibin.calendarview.MonthView;

public class IndexMonthView extends MonthView {

    private Paint mSchemeBasicPaint = new Paint();
    private int mPadding;
    private int mH, mW;
    private int mRadius;

    public IndexMonthView(Context context) {
        super(context);

        mSchemeBasicPaint.setAntiAlias(true);
        mSchemeBasicPaint.setStyle(Paint.Style.FILL);
        mSchemeBasicPaint.setTextAlign(Paint.Align.CENTER);
        mSchemeBasicPaint.setColor(0xff333333);
        mSchemeBasicPaint.setFakeBoldText(true);
        mPadding = CalendarUtil.dipToPx(getContext(), 4);
        mH = CalendarUtil.dipToPx(getContext(), 2);
        mW = CalendarUtil.dipToPx(getContext(), 8);
    }

    @Override
    protected void onPreviewHook() {
        mRadius = Math.min(mItemWidth, mItemHeight) * 5 / 12;
    }


    @Override
    protected boolean onDrawSelected(Canvas canvas, Calendar calendar, int x, int y, boolean hasScheme) {
//        mSelectedPaint.setStyle(Paint.Style.FILL);
//        canvas.drawRect(x + mPadding, y + mPadding, x + mItemWidth - mPadding, y + mItemHeight - mPadding, mSelectedPaint);
//        return true;
        int cx = x + mItemWidth / 2;
        int cy = y + mItemHeight / 2;
        canvas.drawCircle(cx, cy, mRadius, mSelectedPaint);
        return true;
    }

    /**
     * onDrawSelected
     * @param canvas   canvas
     * @param calendar 日历calendar
     * @param x        日历Card x起点坐标
     * @param y        日历Card y起点坐标
     */
    @SuppressWarnings("IntegerDivisionInFloatingPointContext")
    @Override
    protected void onDrawScheme(Canvas canvas, Calendar calendar, int x, int y) {
        mSchemeBasicPaint.setColor(calendar.getSchemeColor());
        canvas.drawRect(x + mItemWidth / 2 - mW / 2,
                y + mItemHeight - mH * 2 - mPadding,
                x + mItemWidth / 2 + mW / 2,
                y + mItemHeight - mH - mPadding, mSchemeBasicPaint);
    }

    @SuppressWarnings("IntegerDivisionInFloatingPointContext")
    @Override
    protected void onDrawText(Canvas canvas, Calendar calendar, int x, int y, boolean hasScheme, boolean isSelected) {
        int cx = x + mItemWidth / 2;
        int top = y - mItemHeight / 6;
        if (hasScheme) {
            canvas.drawText(String.valueOf(calendar.getDay()), cx, mTextBaseLine + top,
                    calendar.isCurrentDay() ? mCurDayTextPaint :
                            calendar.isCurrentMonth() ? mSchemeTextPaint : mOtherMonthTextPaint);

            canvas.drawText(calendar.getLunar(), cx, mTextBaseLine + y + mItemHeight / 10,
                    calendar.isCurrentDay() ? mCurDayLunarTextPaint :
                            mCurMonthLunarTextPaint);

        } else {
            canvas.drawText(String.valueOf(calendar.getDay()), cx, mTextBaseLine + top,
                    calendar.isCurrentDay() ? mCurDayTextPaint :
                            calendar.isCurrentMonth() ? mCurMonthTextPaint : mOtherMonthTextPaint);
            canvas.drawText(calendar.getLunar(), cx, mTextBaseLine + y + mItemHeight / 10, mCurMonthLunarTextPaint);
        }
    }

}

File: android/app/src/main/java/com/gang/notodo/ui/calendar/IndexMonthView.kt
package com.gang.notodo.ui.calendar


import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint

import com.gang.notodo.util.CalendarUtil
import com.haibin.calendarview.Calendar
import com.haibin.calendarview.MonthView
import kotlin.math.min


class IndexMonthView(context: Context) : MonthView(context) {

    private val mPadding = CalendarUtil.dipToPx(context, 4f)
    private val mH = CalendarUtil.dipToPx(context, 2f)
    private val mW = CalendarUtil.dipToPx(context, 8f)
    private var mRadius: Int = 0

    private val mSchemeBasicPaint = Paint().apply {
        this.isAntiAlias = true
        this.style = Paint.Style.FILL
        this.textAlign = Paint.Align.CENTER
        this.color = -0xcccccd
        this.isFakeBoldText = true
    }


    override fun onPreviewHook() {
        mRadius = min(mItemWidth, mItemHeight) * 5 / 12
    }


    override fun onDrawSelected(
        canvas: Canvas,
        calendar: Calendar,
        x: Int,
        y: Int,
        hasScheme: Boolean
    ): Boolean {
        //        mSelectedPaint.setStyle(Paint.Style.FILL);
        //        canvas.drawRect(x + mPadding, y + mPadding, x + mItemWidth - mPadding, y + mItemHeight - mPadding, mSelectedPaint);
        //        return true;
        val cx = x + mItemWidth / 2
        val cy = y + mItemHeight / 2
        canvas.drawCircle(cx.toFloat(), cy.toFloat(), mRadius.toFloat(), mSelectedPaint)
        return true
    }

    /**
     * onDrawSelected
     *
     * @param canvas   canvas
     * @param calendar 日历calendar
     * @param x        日历Card x起点坐标
     * @param y        日历Card y起点坐标
     */
    override fun onDrawScheme(canvas: Canvas, calendar: Calendar, x: Int, y: Int) {
        mSchemeBasicPaint.color = calendar.schemeColor
        canvas.drawRect(
            (x + mItemWidth / 2 - mW / 2).toFloat(),
            (y + mItemHeight - mH * 2 - mPadding).toFloat(),
            (x + mItemWidth / 2 + mW / 2).toFloat(),
            (y + mItemHeight - mH - mPadding).toFloat(), mSchemeBasicPaint
        )
    }

    override fun onDrawText(
        canvas: Canvas,
        calendar: Calendar,
        x: Int,
        y: Int,
        hasScheme: Boolean,
        isSelected: Boolean
    ) {
        val cx = x + mItemWidth / 2
        val top = y - mItemHeight / 6
        if (hasScheme) {
            canvas.drawText(
                calendar.day.toString(), cx.toFloat(), mTextBaseLine + top,
                when {
                    calendar.isCurrentDay -> mCurDayTextPaint
                    calendar.isCurrentMonth -> mSchemeTextPaint
                    else -> mOtherMonthTextPaint
                }
            )

            canvas.drawText(
                calendar.lunar,
                cx.toFloat(),
                mTextBaseLine + y.toFloat() + (mItemHeight / 10).toFloat(),
                if (calendar.isCurrentDay)
                    mCurDayLunarTextPaint
                else
                    mCurMonthLunarTextPaint
            )

        } else {
            canvas.drawText(
                calendar.day.toString(), cx.toFloat(), mTextBaseLine + top,
                when {
                    calendar.isCurrentDay -> mCurDayTextPaint
                    calendar.isCurrentMonth -> mCurMonthTextPaint
                    else -> mOtherMonthTextPaint
                }
            )
            canvas.drawText(
                calendar.lunar,
                cx.toFloat(),
                mTextBaseLine + y.toFloat() + (mItemHeight / 10).toFloat(),
                mCurMonthLunarTextPaint
            )
        }
    }

}
